internal/configfsm/fsm.go

```diff
@@ -0,0 +1,315 @@
+	aState, err = fsm.State(ctx)
+	if err != nil {
+		return aState, fmt.Errorf("could not get state, err: %w", err)
+	}
+	switch aState {
+	case Start:
+		err = fsm.Fire(StartToMissing, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case Missing:
+		err = fsm.Fire(MissingToMissing, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case ClusterNotReady:
+		err = fsm.Fire(ClusterNotReadyToClusterNotReady, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case InProgress:
+		err = fsm.Fire(InProgressToInProgress, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case TimedOut:
+		err = fsm.Fire(TimedOutToTimedOut, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case Completed:
+		err = fsm.Fire(CompletedToCompleted, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	case OutOfDate:
+		err = fsm.Fire(OutOfDateToOutOfDate, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	}
+	// Update the current state
+	aState, err = fsm.State(ctx)
+	if err != nil {
+		return aState, fmt.Errorf("could not get state, err: %w", err)
+	}
+	return aState, nil
+}
+
+// Initialize the state machine algorithm
+func InitFSM(state string) (fsm *stateless.StateMachine, err error) {
+	fsm = stateless.NewStateMachine(state)
+	fsm.Configure(Start).Permit(StartToMissing, Missing)
+	fsm.Configure(ClusterNotReady).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering ClusterNotReady")
+			if fsmHelper.IsTimedOut() {
+				return fsm.Fire(ClusterNotReadyToTimedOut, fsmHelper)
+			}
+			if fsmHelper.IsClusterReady() {
+				if fsmHelper.IsNonCompliantPolicyInEnforce() || fsmHelper.IsAllPoliciesCompliant() {
+					return fsm.Fire(ClusterNotReadyToInProgress, fsmHelper)
+				} else {
+					return fsm.Fire(ClusterNotReadyToOutOfDate, fsmHelper)
+				}
+			}
+			return nil
+		}).
+		Permit(ClusterNotReadyToMissing, Missing).
+		Permit(ClusterNotReadyToInProgress, InProgress).
+		Permit(ClusterNotReadyToTimedOut, TimedOut).
+		Permit(ClusterNotReadyToOutOfDate, OutOfDate).
+		PermitReentry(ClusterNotReadyToClusterNotReady)
+	fsm.Configure(InProgress).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering InProgress")
+			if fsmHelper.IsResetNonCompliantAt() {
+				fsmHelper.SetResetNonCompliantAtNow()
+			}
+
+			if fsmHelper.IsTimedOut() {
+				return fsm.Fire(InProgressToTimedOut, fsmHelper)
+			}
+			if fsmHelper.IsAllPoliciesCompliant() {
+				return fsm.Fire(InProgressToCompleted, fsmHelper)
```

Comment:
If `Completed` is the final state then its pre-conditions should be checked last so that if any of the other conditions are true they take precedence.

Reply:
Fixed.